lib/include/openamp/virtio_mmio.h

```diff
@@ -301,6 +336,96 @@ int virtio_mmio_device_init(struct virtio_mmio_device *vmdev, uintptr_t virt_mem
 
 void virtio_mmio_isr(struct virtio_device *vdev);
 
+#if defined(HVL_VIRTIO)
+
+/**
+ * @brief VIRTIO MMIO shared memory pool initialization routine.
+ *
+ * @param[in] mem Pointer to memory.
+ * @param[in] size Size of memory region in bytes.
+ *
+ * @return N/A.
+ */
+
+void virtio_mmio_shm_pool_init(void *mem, size_t size);
+
+/**
+ * @brief VIRTIO MMIO shared memory pool buffer allocation routine.
+ *
+ * @param[in] size Number of bytes requested.
+ *
+ * @return pointer to allocated memory space in shared memory region.
+ */
+
+void *virtio_mmio_shm_pool_alloc(size_t size);
+
+/**
+ * @brief VIRTIO MMIO shared memory pool buffer deallocation routine.
+ *
+ * @param[in] ptr Pointer to memory space to free.
+ *
+ * @return N/A.
+ */
+
+void virtio_mmio_shm_pool_free(void *ptr);
+
+/**
+ * @brief VIRTIO MMIO (hypervisorless mode) inter-processor notification routine.
+ *
+ * @return N/A.
+ */
+
+void virtio_mmio_hvl_ipi(void);
+
+/**
+ * @brief VIRTIO MMIO (hypervisorless mode) wait routine.
+ *
+ * @param[in] usec Number of microseconds to wait.
+ *
+ * @return N/A.
+ */
+
+void virtio_mmio_hvl_wait(uint32_t usec);
+
+/**
+ * @brief Add VIRTIO MMIO (hypervisorless mode) callback.
+ *
+ * @param[in] func Callback function pointer.
+ * @param[in] arg  Callback parameter.
+ *
+ * @return int 0 for success.
+ */
+
+int virtio_mmio_hvl_cb_set(virtio_mmio_hvl_cb_t func, void *arg);
+
+/**
+ * @brief VIRTIO MMIO (hypervisorless mode) callback execution routine.
+ *
+ * @return N/A
+ */
+
+void virtio_mmio_hvl_cb_run(void);
+
+/**
+ * @brief Add VIRTIO MMIO (hypervisorless mode) shared-memory bounce buffer.
+ *
+ * @return bounce buffer physical address or METAL_BAD_PHYS
+ */
+
+uint64_t virtio_mmio_hvl_add_bounce_buf(struct virtqueue *vq, void *cookie, char *buffer,
+					unsigned int len);
+
+/**
+ * @brief Get used VIRTIO MMIO (hypervisorless mode) shared-memory bounce buffer.
+ *
+ * @return N/A
+ */
+
+void virtio_mmio_hvl_get_bounce_buf(struct virtqueue *vq, struct vring_desc *dp,
+				    void *cookie);
+
+#endif /* defined(HVL_VIRTIO) */
+
```

Comment:
What about creating a virtio_mmio_hvl.c to split generic virtio_mmio and hypervisorless services?

Reply:
Done in the latest version of the PR.